### PR TITLE
Add json response body for API 401 cases

### PR DESCRIPTION
**Pull request by sahandilshan**
### Proposed changes in this pull request

Fixed https://github.com/wso2/product-is/issues/12830 by adding a proper JSON response.

Screenshot 2021-11-29 at 20 11 16

Screenshot 2021-11-29 at 20 11 32

### CODE REVIEW

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/AuthenticationValve.java [hunk 1]
```diff
@@ -243,7 +244,14 @@ private void handleErrorResponse(AuthenticationContext authenticationContext, Re
         }
         value.append('\"');
         response.setHeader(AUTH_HEADER_NAME, value.toString());
-        response.sendError(error);
+        response.setStatus(error);
+        response.setHeader("Content-Type","application/json;charset=UTF-8");
+        JSONObject jsonObject = new JSONObject();
+        jsonObject.put("message", "Requires authentication");
+        jsonObject.put("documentation_url", "https://docs.wso2.com/display/IS511/apidocs/Authentication-framework-"
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/AuthenticationValve.java:251]: Can we rethink adding the documentation link here?
- In this Authentication valve layer, we can't specifically return doc related to invoked API. 
- If we are pointing to the documentation, then we need to point to the correct API doc version
- What If other solutions have their own documentation pages?

**Comment by senthalan** [comment]
Can check include error message for response code returned as well ?

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/AuthenticationValve.java (hunk 1 above)
Comment (sahandilshan) [reply components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/AuthenticationValve.java]: This has been resolved with this commit https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/164/commits/d7bca3033297e9d011833946b74a9623db5311f5

Followed the standard error response format mentioned [here](https://datatracker.ietf.org/doc/html/rfc7644#section-3.7.3). This is the same error response format mentioned in the IS docs as well for the SCIM API

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
+package org.wso2.carbon.identity.auth.valve.util;
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:1]: License header missing

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.wso2.carbon.identity.auth.service.AuthenticationContext;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+import org.apache.log4j.MDC;
+
+public class APIErrorResponseHandler {
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:15]: let's add class level doc comment

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    }
+
+
+    private static boolean isRequestFromScim2Api(Response response) {
+
+        if (response.getRequest() == null) {
+            return false;
+        }
+        return response.getRequest().getRequestURI().contains("/scim2");
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:50]: can we check starts with ? that will be more accurate

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+import org.apache.log4j.MDC;
+
+public class APIErrorResponseHandler {
+
+    private static final Log log = LogFactory.getLog(APIErrorResponseHandler.class);
+
+    private static final String AUTH_HEADER_NAME = "WWW-Authenticate";
+    private static final String CORRELATION_ID_MDC = "Correlation-ID";
+
+    public static void handleErrorResponse(AuthenticationContext authenticationContext, Response response, int error,
```
Comment (senthalan) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:22]: Doc comment is needed for all the public methods

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java (hunk 2 above)
Comment (sahandilshan) [reply components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:17]: This has been fixed with 747c2c448717dcac5a09b0abec9838c5d61ac644

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java (hunk 3 above)
Comment (sahandilshan) [reply components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:34]: This has been fixed with 747c2c448717dcac5a09b0abec9838c5d61ac644

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java (hunk 4 above)
Comment (sahandilshan) [reply components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java]: This has been fixed with 747c2c448717dcac5a09b0abec9838c5d61ac644

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java (hunk 5 above)
Comment (sahandilshan) [reply components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:45]: This has been fixed with 747c2c448717dcac5a09b0abec9838c5d61ac644

**Review by senthalan — APPROVED** [review]

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.wso2.carbon.identity.auth.service.AuthenticationContext;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+import org.apache.log4j.MDC;
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:29]: Let's try to use slf4j library

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+
+    private static boolean isRequestFromScim2Api(String uri) {
+
+        if (uri == null) {
+            return false;
+        }
+        return uri.startsWith("/scim2");
+    }
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:100]: Let's simplify this.
eg: `return StringUtils.isNotBlank(uri) && uri.startsWith("/scim2")`

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+    private static boolean isRequestFromDCREndpoint(String uri) {
+        if (uri == null) {
+            return false;
+        }
+        return uri.startsWith("/api/identity/oauth2/dcr");
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:106]: Let's simplify this

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    private static void handleScim2ApiErrorResponse(Response response, int error, Exception e) throws IOException {
+
+        response.setStatus(error);
+        response.setHeader("Content-Type","application/json;charset=UTF-8");
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:112]: We should have constants for Strings like "content-Type". Let's reuse them

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+            jsonObject.put("description", errorDescription);
+        }
+        if (isCorrelationIDPresent()) {
+            jsonObject.put("traceId", MDC.get(CORRELATION_ID_MDC));
+        }
+        setResponseBody(response, jsonObject);
+    }
+
+    private static void setResponseBody(Response response, JSONObject jsonObject) throws IOException {
+        response.getWriter().write(jsonObject.toString(2));
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:202]: add blank line here

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+        response.finishResponse();
+    }
+
+    private static boolean isCorrelationIDPresent() {
+        return MDC.get(CORRELATION_ID_MDC) != null;
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:207]: let's add a blank line after method signature

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java (hunk 6 above)
Comment (sahandilshan) [reply components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:29]: This has already been fixed with https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/175

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java (hunk 11 above)
Comment (sahandilshan) [reply components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:207]: Fixed with https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/183

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java (hunk 10 above)
Comment (sahandilshan) [reply components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:202]: Fixed with https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/183

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java (hunk 9 above)
Comment (sahandilshan) [reply components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:112]: I couldn't find any constant inside the project files for this

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java (hunk 7 above)
Comment (sahandilshan) [reply components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/util/APIErrorResponseHandler.java:100]: Fixed with https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/183

**Pull request merged**